server/commands/connected_objects.go

```diff
@@ -0,0 +1,351 @@
+		}
+	}
+	iter1 := storage.NewStaticTupleIterator(tuples)
+
+	iter2, err := c.Datastore.ReadStartingWithUser(ctx, store, storage.ReadStartingWithUserFilter{
+		ObjectType: req.ingress.Ingress.GetType(),
+		Relation:   tuplesetRelation,
+		UserFilter: []*openfgapb.ObjectRelation{
+			{Object: req.targetUserRef.Object},
+			{Object: "*"},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	iter := storage.NewCombinedIterator(iter1, iter2)
+
+	g := errgroup.Group{}
+	g.SetLimit(100) // set some concurrency limit
+
+	for {
+		t, err := iter.Next()
+		if err != nil {
+			if errors.Is(err, storage.ErrIteratorDone) {
+				break
+			}
+		}
+
+		tk := t.GetKey()
+
+		foundObject := tk.GetObject()
+		foundObjectType, _ := tuple.SplitObject(foundObject)
+
+		if _, ok := foundObjectsMap.Load(foundObject); ok {
+			// todo(jon-whit): we could optimize this by avoiding reading this
+			// from the database in the first place
+
+			// if we've already evaluated/found the object, then continue
+			continue
+		}
+
+		if foundObjectType == sourceObjectType {
+			if foundCount != nil && atomic.LoadUint32(foundCount) >= req.limit {
+				break
+			}
+
+			if foundCount != nil {
+				atomic.AddUint32(foundCount, 1)
+			}
+
+			resultChan <- foundObject
+			foundObjectsMap.Store(foundObject, foundObject)
+		}
+
+		g.Go(func() error {
+			return c.streamedConnectedObjects(ctx, &ConnectedObjectsRequest{
+				StoreID:          store,
+				ObjectType:       sourceObjectType,
+				Relation:         sourceObjectRel,
+				User:             &openfgapb.ObjectRelation{Object: foundObject, Relation: req.targetUserRef.Relation},
+				ContextualTuples: req.contextualTuples,
+			}, resultChan, foundObjectsMap, foundCount)
+		})
+	}
+
+	return g.Wait()
+}
+
+func (c *ConnectedObjectsCommand) reverseExpandDirect(
+	ctx context.Context,
+	req *reverseExpandRequest,
+	resultChan chan<- string,
+	foundObjectsMap *sync.Map,
+	foundCount *uint32,
+) error {
+
+	store := req.storeID
+
+	ingress := req.ingress.Ingress
+
+	sourceObjectType := req.sourceObjectRef.GetType()
+	sourceObjectRel := req.sourceObjectRef.GetRelation()
+
+	var tuples []*openfgapb.Tuple
+	for _, t := range req.contextualTuples {
+
+		object := t.GetObject()
+		objectType, _ := tuple.SplitObject(object)
+		if objectType != ingress.GetType() {
+			continue
+		}
+
+		if t.GetRelation() != ingress.GetRelation() {
+			continue
+		}
+
+		targetUserStr := req.targetUserRef.GetObject()
+		if req.targetUserRef.GetRelation() != "" {
+			targetUserStr = fmt.Sprintf("%s#%s", targetUserStr, req.targetUserRef.GetRelation())
+		}
+
+		if t.GetUser() == targetUserStr || t.GetUser() == "*" {
+			tuples = append(tuples, &openfgapb.Tuple{Key: t})
+		}
+	}
+	iter1 := storage.NewStaticTupleIterator(tuples)
+
+	iter2, err := c.Datastore.ReadStartingWithUser(ctx, store, storage.ReadStartingWithUserFilter{
+		ObjectType: ingress.GetType(),
+		Relation:   ingress.GetRelation(),
+		UserFilter: []*openfgapb.ObjectRelation{
+			req.targetUserRef,
```

Comment:
Could we maybe be explicit here similar to above?

Reply:
I'm not sure I understand what you mean? Could you please clarify?

Reply:
`{Object: req.targetUserRef.Object},`

Reply:
Oh, gotcha! Keyed literals is what you are after. Thanks for the clarification. I'll make that change.

Reply:
Oh, this is intentional. In the `expandDirectUserset` case you reverse read the whole `targetUserRef` (which includes `object#relation`). In the `expandTTU` case you only reverse read the tupleset relation on the `targetUserRef.Object`. 

So these cases are different purposefully.